python/gps/algorithm/algorithm_mdgps.py

```diff
@@ -0,0 +1,341 @@
+            self._hyperparams['policy_opt'], self.dO, self.dU
+        )
+
+        # extra stuff to pass parameters to sampling
+        # TODO: this is sloppy, handle better
+        if not self._hyperparams['agent_use_nn_policy']:
+            del self._hyperparams['agent_use_nn_policy']
+
+    def iteration(self, sample_lists):
+        """
+        Run iteration of BADMM-based guided policy search.
+
+        Args:
+            sample_lists: List of SampleList objects for each condition.
+        """
+        for m in range(self.M):
+            self.cur[m].sample_list = sample_lists[m]
+
+        self._update_dynamics()  # Update dynamics model using all sample.
+        self._update_policy_samples()  # Choose samples to use with the policy.
+        self._update_step_size()  # KL Divergence step size.
+
+        # Run inner loop to compute new policies.
+        for inner_itr in range(self._hyperparams['inner_iterations']):
+            #TODO: Could start from init controller.
+            if self.iteration_count > 0 or inner_itr > 0:
+                # Update the policy.
+                self._update_policy(self.iteration_count, inner_itr)
+            for m in range(self.M):
+                self._update_policy_fit(m)  # Update policy priors.
+
+            # TODO: not sure about this, copied from previous
+            if (self.iteration_count > 0 or inner_itr > 0) and \
+                    (inner_itr == self._hyperparams['inner_iterations'] - 1):
+                for m in range(self.M):
+                    kl_m = self._policy_kl(m)[0]
+                    self.cur[m].pol_info.prev_kl = kl_m
+
+            self._update_trajectories()
+
+        self._advance_iteration_variables()
+
+    def _update_policy_samples(self):
+        """ Update the list of samples to use with the policy. """
+        #TODO: Handle synthetic samples.
+        max_policy_samples = self._hyperparams['max_policy_samples']
+        if self._hyperparams['policy_sample_mode'] == 'add':
+            for m in range(self.M):
+                samples = self.cur[m].pol_info.policy_samples
+                samples.extend(self.cur[m].sample_list)
+                if len(samples) > max_policy_samples:
+                    start = len(samples) - max_policy_samples
+                    self.cur[m].pol_info.policy_samples = samples[start:]
+        else:
+            for m in range(self.M):
+                self.cur[m].pol_info.policy_samples = self.cur[m].sample_list
+
+    def _update_step_size(self):
+        """ Evaluate costs on samples, and adjust the step size. """
+        # Evaluate cost function for all conditions and samples.
+        for m in range(self.M):
+            self._update_policy_fit(m, init=True)
+            self._eval_cost(m)
+            lmax = self.cur[m].cs.max(axis=0)
+            lmin = self.cur[m].cs.min(axis=0)
+            self.cur[m].qmax = np.cumsum( (lmax - lmin)[::-1] )[::-1]
```

Comment:
This was for an idea Sergey had where we compute the cumulative Qmax (like in the bound from the paper), in order to try and inform the step size updates. We never got it working, and I forgot to remove it.